### ISSUE TITLE
Numpy support for ray tune config | numpy array is not JSON serializable

### ISSUE THREAD

**Issue by timediv** [ray-project/ray#2243]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Docker nvidia/cuda:9.0-base-ubuntu16.04
- **Ray installed from (source or binary)**: Binary
- **Ray version**: ray==0.4.0
- **Python version**: Python 3.5.2
- **Exact command to reproduce**:
Ray tune config with numpy arrays:
`
'config': {'var:' np.array([1.0, 2.0])}
`

### Describe the problem
Ray tune tries to convert the numpy array to JSON and fails.

```
Traceback (most recent call last):
  File "experiments.py", line 245, in <module>
    tune.run_experiments({'pbt_modularity': train_spec}, scheduler=pbt)
  File "/usr/local/lib/python3.5/dist-packages/ray/tune/tune.py", line 63, in run_experiments
    for trial in generate_trials(spec, name):
  File "/usr/local/lib/python3.5/dist-packages/ray/tune/variant_generator.py", line 48, in generate_trials
    args = parser.parse_args(to_argv(spec))
  File "/usr/local/lib/python3.5/dist-packages/ray/tune/variant_generator.py", line 20, in to_argv
    argv.append(json.dumps(v))
  File "/usr/lib/python3.5/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.5/json/encoder.py", line 198, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.5/json/encoder.py", line 256, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.5/json/encoder.py", line 179, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: array([1.61153255e-05, 2.11210111e-05, 9.83538792e-06]) is not JSON serializable
```

The conversion to JSON happens here (variant_generator.py line 41 and following):
```
            try:
                # Special case the `env` param for RLlib by automatically
                # moving it into the `config` section.
                if "env" in spec:
                    spec["config"] = spec.get("config", {})
                    spec["config"]["env"] = spec["env"]
                    del spec["env"]
                args = parser.parse_args(to_argv(spec))
            except SystemExit:
                raise TuneError("Error parsing args, see above message", spec)
            if resolved_vars:
                experiment_tag = "{}_{}".format(i, resolved_vars)
            else:
                experiment_tag = str(i)
            i += 1
            yield Trial(
                trainable_name=spec["run"],
                config=spec.get("config", {}),
                local_dir=os.path.join(args.local_dir, output_path),
                experiment_tag=experiment_tag,
                resources=json_to_resources(spec.get("trial_resources", {})),
                stopping_criterion=spec.get("stop", {}),
                checkpoint_freq=args.checkpoint_freq,
                restore_path=spec.get("restore"),
                upload_dir=args.upload_dir,
                max_failures=args.max_failures)
```

Relevant line: `  args = parser.parse_args(to_argv(spec))`

I don't see why this conversion is necessary in the first place.
Because there are only few fields of the spec (args.upload_dir, args.max_failures, ...) that are read this way and all of them could be directly obtained from the spec object?

**Comment by ericl** [comment]
Hey Louis, while it is possible to support picklable objects as well, JSON format is convenient for generating portable logs.

As a workaround, you should be able to use tolist() / np.ndarray() to convert to / from json serializable objects. If the array is particularly large, and you don't need it to be logged, you can also consider broadcasting it via the object store.

**Comment by richardliaw** [comment]
Hi @timediv, would it be possible to give a bit more context as to what you're trying to do?

**Comment by ericl** [comment]
I see, so you want an array type to be mutated by PBT? In that case it probably makes sense to use `custom_explore_fn`, and then you don't need to put the list in the mutation params.

**Comment by timediv** [comment]
@richardliaw My goal is to have a scalar hyperparameter per layer in a neural network of depth `D`. I was hoping I could just create a numpy array of shape `D` as a hyperparameter in `config`.
Then I can specify a distribution `lambda _: np.random.uniform(x, y, size=D)` as well as a pertubation function for PBT in the form of `lambda: np.random.uniform(a, b, size=D)`.
I noticed that numpy arrays are not supported as a value for entries in `config` because it raises the above error.
Of course, I could just create `D` keys in the form of a dict `{f'layer{d}/param': func for d in range(D)` but this is not very nice to handle in my opinion.
And also doesn't generalize well if we'd like to have say hyperparameters of shape `[X, Y, Z]`.

@ericl The problem is that even `config` does not support numpy arrays. To me it seems like the mutation code of PBT should work just fine with numpy arrays.

**Comment by ericl** [comment]
I see in this case converting to a list will work fine. Tune doesn't treat
lists as categorical variables unless you wrap them in tune.grid_search(),
so you should be good there.

You can then use a custom explore fn to apply a custom mutation on the list.

**Comment by richardliaw** [comment]
Hey @timediv did Eric's suggestion resolve the issue?

**Comment by ericl** [comment]
https://github.com/ray-project/ray/pull/2287 should fix this

**Comment by timediv** [comment]
Amazing! Will try it later. That will result in much cleaner code, thanks!

**Comment by ericl** [comment]
Should be fixed by #2287
